tuf/repository_tool.py

```diff
@@ -2530,14 +2546,58 @@ def delegate_hashed_bins(self, list_of_targets, keys_of_hashed_bins,
       hash_prefix = _get_hash(target_path.replace('\\', '/').lstrip('/'))[:prefix_length]
       ordered_roles[int(hash_prefix, 16) // bin_size]["target_paths"].append(target_path)
 
+    keyids, keydict = _keys_to_keydict(keys_of_hashed_bins)
+
+    # A queue of roleinfo's that need to be updated in the roledb
+    delegated_roleinfos = []
+
     for bin_rolename in ordered_roles:
+      # TODO: originally we just called self.delegate() for each item in this
+      # iteration. However, this is *extremely* slow when creating a large
+      # number of hashed bins, i.e. 16k as is recommended for PyPI usage in
+      # PEP 458: https://www.python.org/dev/peps/pep-0458/
+      # The source of the slowness is the interactions with the roledb, which
+      # causes several deep copies of roleinfo dictionaries:
+      # https://github.com/theupdateframework/tuf/issues/1005
+      # Once the underlying issues in #1005 are resolved, i.e. some combination
+      # of the intermediate and long-term fixes, we may simplify here by
+      # switching back to just calling self.delegate(), but until that time we
+      # queue roledb interactions and perform all updates to the roledb in one
+      # operation at the end of the iteration.
+
+      relative_paths = {}
+      targets_directory_length = len(self._targets_directory)
+      for path in bin_rolename['target_paths']:
+        relative_paths.update({path[targets_directory_length:]: {}})
+
       # Delegate from the "unclaimed" targets role to each 'bin_rolename'
-      self.delegate(bin_rolename['name'], keys_of_hashed_bins, [],
-          list_of_targets=bin_rolename['target_paths'],
-          path_hash_prefixes=bin_rolename['target_hash_prefixes'])
+      target = self._create_delegated_target(bin_rolename['name'], keyids,
+          paths=relative_paths)
```

Comment:
May I suggest to pass the same hardcoded threshold of `1` as below? This is what `delegate()` would have done.
https://github.com/theupdateframework/tuf/blob/32326975e040ad82e22d66ec27823a6122e49136/tuf/repository_tool.py#L2318
https://github.com/theupdateframework/tuf/blob/32326975e040ad82e22d66ec27823a6122e49136/tuf/repository_tool.py#L2336

As a consequence we could  remove the default for threshold in `_create_delegated_target`. What do you think?

Reply:
Squashed in to 6703b51